### PR TITLE
Launcher now reaps orphaned zombie processes

### LINKED ISSUE #4412: Supervisor/Launcher should reap zombie processes
Sometimes parents die and leave their children orphaned. When this happens in the Habitat supervision tree, the process typically ends up owned by the launcher. If the launcher is owned by something like systemd, the zombies usually get reaped. However, in the case where the launcher is PID 1 the zombies live on for ever and eat your containers BRAINZ (causing it to crash). The launcher/supervisor should detect when it has zombied processes in it's tree and proactively reap them.

### PR DESCRIPTION
**Pull request by christophermaier**
tenor-4706015
tenor-10240249
tenor-25218364

For details, read the commit messages.

⚠️ Alert! ⚠️ 
Requires habitat-sh/core#9 to merge first! DO NOT MERGE BEFORE THAT! Until then, if you want to test this locally, follow the instructions in the second commit; sorry.

Closes #4412

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/launcher/src/server/mod.rs [hunk 1]
```diff
@@ -153,17 +172,136 @@ impl Server {
...
+                Some(self.handle_supervisor_exit(None))
+            }
+        } else {
+            // A supervisor didn't end; carry on your merry way
+            None
+        }
+    }
+
+    /// Windows doesn't have the same orphan-reaping behavior as Linux;
+    /// returnig `None` means that there's nothing special that needs
```
Comment (raskchanky) [inline components/launcher/src/server/mod.rs]: `returnig` -> `returning`

**Comment by raskchanky** [comment]
I don't know enough about linux signal handling to say whether this implementation is robust or not, but I did find a spelling error 😁 and I learned a lot reading through it.

File: components/launcher/src/server/mod.rs [hunk 2]
```diff
@@ -153,17 +172,136 @@ impl Server {
...
+    ///
+    /// Note that we are not (yet?) doing anything with
+    /// `prctl(PR_SET_CHILD_SUBREAPER, ...)` to make the Launcher a
+    /// subreaper; this behavior currently handles the case when the
+    /// Launcher is running as PID 1.
+    ///
+    /// (See http://man7.org/linux/man-pages/man2/prctl.2.html for
+    /// further information.)
+    #[cfg(unix)]
+    fn reap_zombie_orphans(&mut self) -> Option<Result<TickState>> {
```
Comment (reset) [inline components/launcher/src/server/mod.rs]: Yup, this is something I should have caught when I wrote this code. Looks perfect to me!

**Comment by reset** [comment]
tenor-44048280

@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/347438036).

**Comment by thesentinels** [comment]
:broken_heart: Travis CI reports [this PR failed to pass the test suite](https://travis-ci.org/habitat-sh/habitat/builds/347438036).

The next step is to examine the [job](https://travis-ci.org/habitat-sh/habitat/builds/347438036) and figure out why. If it is transient, you can try re-triggering the Travis CI Job - if it passes, this PR will be automatically merged. If it is not transient, you should fix the issue and update this pull request, and issue `approve` again. If you believe it will never pass, and you are feeling :godmode:, you can issue a `force` to merge this PR anyway.

File: Cargo.lock [hunk 3]
```diff
@@ -487,7 +487,7 @@ dependencies = [
  "env_logger 0.5.3 (registry+https://github.com/rust-lang/crates.io-index)",
  "habitat_api_client 0.0.0",
  "habitat_common 0.0.0",
- "habitat_core 0.0.0 (git+https://github.com/habitat-sh/core.git)",
```
Comment (fnichol) [inline Cargo.lock]: I'm wondering if this change was due to temporarily overriding the source of `core`? Is it possible this isn't intended?

File: Cargo.toml [hunk 4]
```diff
@@ -24,3 +24,7 @@ members = [
   "components/segment-api-client",
   "components/sup"
 ]
+
+[patch."https://github.com/habitat-sh/core.git"]
```
Comment (fnichol) [inline Cargo.toml]: Ah yeah, I'm guessing this shouldn't be in the PR…

File: Cargo.lock (hunk 3 above)
Comment (mwrock) [reply Cargo.lock]: yeah this definitely looks like a dev hack not meant for the world.

**Comment by christophermaier** [comment]
@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/347511479).

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/347511479).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

**Pull request merged**